src/shim.c

```diff
@@ -703,6 +704,9 @@ handle_proxy_message(struct cc_shim *shim)
 
 	fr = read_frame(shim);
 	if ( !fr) {
+		if (! reconnect_to_proxy(shim)) {
```

Comment:
Sorry for the delayed response. 
So read_frame() can return NULL in case when read_wire_data() fails --> we re-connect.
Also read_frame() return NULL in cases:
 * shim == NULL or shim->proxy_sock_fd < 0
 * fr->header.header_len < MIN_HEADER_WORD_SIZE
 * fr->header.version != PROXY_VERSION

My proposition is to handle these cases the same as failures with allocation --> calling abort(). Doing this way will make read_frame() return NULL only in cases when read_wire_date() failed.